### PR TITLE
lesson01_assignment | lesson02_assignment | lesson03_assignment | lesson04_assignment

### PR DESCRIPTION
**Pull request by visokoo**
See the `README.md` under `tests/` for instructions on how to run tests.

- Added .gitignore
- Edited `assignment/` `README.md` to include link to `tests/` `README.md`
- Added edited class files under `./inventory_management`
- Added `integration_test.py` and `unit_tests.py` under `tests/`
- Added lesson02 files
- Added lesson03 files
- Added lesson04 files

### CODE REVIEW

File: students/visokoo/lesson01/assignment/tests/integration_test.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+
+class ModuleTests(TestCase):
+    """ Combining all the classes and testing collectively """
+
+    def test_main(self):
+        """ Testing item creation with all classes and viewing the contents """
+        with patch('builtins.input', lambda value: '1'):
+            main.main_menu()
+            market_prices.get_latest_price = MagicMock(return_value=300)
```
Comment (majstorpoki) [inline students/visokoo/lesson01/assignment/tests/integration_test.py:17]: Why are you testing with 300?

File: students/visokoo/lesson01/assignment/tests/unit_tests.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+class InventoryTests(TestCase):
+    """
+    InventoryTests class to test method functionality
+    """
+    def test_init(self):
+        """
+        Test that a proper Inventory object is returned
+        """
+        item = Inventory(1, "Item1", "200", "400")
```
Comment (majstorpoki) [inline students/visokoo/lesson01/assignment/tests/unit_tests.py:22]: Which part of code are you testing here?

File: students/visokoo/lesson01/assignment/tests/unit_tests.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        self.assertEqual("200", item.market_price)
+        self.assertEqual("400", item.rental_price)
+
+    def test_dict(self):
+        """
+        Test that a proper Inventory dict is returned
+        """
+        item = Inventory(1, "Item1", "200", "400")
+
+        output_dict = item.return_as_dictionary()
```
Comment (majstorpoki) [inline students/visokoo/lesson01/assignment/tests/unit_tests.py:34]: So, you are testing "return_as_dictionary"

File: students/visokoo/lesson01/assignment/tests/unit_tests.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+            {
+                'product_code': 1,
+                'description': 'Item1',
+                'market_price': '200',
+                'rental_price': '400'
+            },
+            output_dict)
+
+
+class ElectricAppliancesTests(TestCase):
```
Comment (majstorpoki) [inline students/visokoo/lesson01/assignment/tests/unit_tests.py:45]: I like that you split them in classes depending on what do you test

File: students/visokoo/lesson01/assignment/tests/unit_tests.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+            },
+            new_item)
+
+    def test_get_info(self):
+        """ Test getting item info """
+        input_info = [1, 'Dryer', '500', 'n', 'y', 'GE', '220']
+        with patch('builtins.input', side_effect=input_info):
+            main.add_new_item()
+        with patch('builtins.input', lambda value: 1):
+            item_dict = main.item_info()[0]
```
Comment (majstorpoki) [inline students/visokoo/lesson01/assignment/tests/unit_tests.py:173]: What this exactly does? Why do you have an index here?

File: students/visokoo/lesson01/assignment/tests/README.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
++-----------+-------+---------+-----------+
+|refactor   |0      |0        |=          |
++-----------+-------+---------+-----------+
+|warning    |0      |0        |=          |
++-----------+-------+---------+-----------+
+|error      |0      |0        |=          |
++-----------+-------+---------+-----------+
+
+--------------------------------------------------------------------
+Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```
Comment (majstorpoki) [inline students/visokoo/lesson01/assignment/tests/README.md:111]: :)

File: students/visokoo/lesson02/assignment/extras-optional/README.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# Terrible program
```
Comment (majstorpoki) [inline students/visokoo/lesson02/assignment/extras-optional/README.md:1]: ??? :)

File: students/visokoo/lesson02/assignment/src/charges_calc.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+import json
+import datetime
+import math
+import logging
+import logging.config
+
+
+def init_logging(logging_level):
+    """ Setting up the logger
+    lvl 1 = only show error msgs on console and in file
```
Comment (majstorpoki) [inline students/visokoo/lesson02/assignment/src/charges_calc.py:14]: I like the comment here

File: students/visokoo/lesson02/assignment/src/charges_calc.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        file_handler.setLevel(logging.ERROR)
+        stream_handler.setLevel(logging.ERROR)
+        logger.addHandler(file_handler)
+        logger.addHandler(stream_handler)
+    elif logging_level == 2:
+        file_handler.setLevel(logging.WARNING)
+        stream_handler.setLevel(logging.WARNING)
+        logger.addHandler(file_handler)
+        logger.addHandler(stream_handler)
+    elif logging_level == 3:
```
Comment (majstorpoki) [inline students/visokoo/lesson02/assignment/src/charges_calc.py:41]: Did you check this functionality?

File: students/visokoo/lesson02/assignment/src/charges_calc.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+                value['total_days'] = (rental_end - rental_start).days
+                value['total_price'] = (
+                    value['total_days'] * value['price_per_day'])
+                value['sqrt_total_price'] = math.sqrt(value['total_price'])
+                value['unit_cost'] = (
+                    value['total_price'] / value['units_rented'])
+                LOGGER.debug("Current item after changes: %s", value)
+        except ValueError as value_error:
+            if value['rental_end'] == "":
+                LOGGER.warning(
```
Comment (majstorpoki) [inline students/visokoo/lesson02/assignment/src/charges_calc.py:108]: I like how you set different logger levels.

File: students/visokoo/lesson01/assignment/tests/integration_test.py (hunk 1 above)
Comment (visokoo) [reply students/visokoo/lesson01/assignment/tests/integration_test.py:17]: Testing a random value to make sure MagicMock is returning what I'm setting and it's validated in the `self.assertEqual` that follows this.

File: students/visokoo/lesson01/assignment/tests/unit_tests.py (hunk 2 above)
Comment (visokoo) [reply students/visokoo/lesson01/assignment/tests/unit_tests.py:22]: This part is testing whether or not the values that I'm passing to the constructor are forming the Inventory object correctly.

File: students/visokoo/lesson01/assignment/tests/unit_tests.py (hunk 3 above)
Comment (visokoo) [reply students/visokoo/lesson01/assignment/tests/unit_tests.py:34]: Yes, my function doc string specifically states that I'm testing for a proper Inventory dict to be returned. Should the doc string be changed to something that's more specific?

File: students/visokoo/lesson01/assignment/tests/unit_tests.py (hunk 5 above)
Comment (visokoo) [reply students/visokoo/lesson01/assignment/tests/unit_tests.py:173]: The function that I'm testing, `def item_info()` returns a list with one dict so the index is pointing to that dict and checking that the values match the input that was passed in.

File: students/visokoo/lesson02/assignment/src/charges_calc.py (hunk 9 above)
Comment (visokoo) [reply students/visokoo/lesson02/assignment/src/charges_calc.py:41]: I did. Works great! I could have done it the same way as above, but wanted to give the .conf file a try. It results in a log file that's name is based on the date format given and only logs messages up to warning while it logs messages all the way up to debug on the console.

File: students/visokoo/lesson02/assignment/extras-optional/README.md (hunk 7 above)
Comment (visokoo) [reply students/visokoo/lesson02/assignment/extras-optional/README.md:1]: This was part original repo, I swear!

File: students/visokoo/lesson03/assignment/src/basic_operations.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
+"""
+    A compilation of methods for interacting with an ORM DATABASE
+"""
+
+import logging
+from playhouse.shortcuts import model_to_dict
```
Comment (majstorpoki) [inline students/visokoo/lesson03/assignment/src/basic_operations.py:6]: What is this module for?

File: students/visokoo/lesson03/assignment/src/basic_operations.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+                name=name,
+                last_name=last_name,
+                home_address=home_address,
+                phone_number=phone_number,
+                email_address=email_address,
+                status=status,
+                credit_limit=credit_limit)
+            LOGGER.info("DATABASE add successful")
+    except peewee.IntegrityError:
+        LOGGER.info("Unique constraint failed on %s", customer_id)
```
Comment (majstorpoki) [inline students/visokoo/lesson03/assignment/src/basic_operations.py:50]: LIKE THIS!

File: students/visokoo/lesson03/assignment/src/basic_operations.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+    :rtype dict
+    """
+    query_dict = {}
+    try:
+        query = (
+            Customer
+            .select()
+            .where(Customer.customer_id == customer_id).get())
+        query_dict = model_to_dict(query)
+    except peewee.DoesNotExist:
```
Comment (majstorpoki) [inline students/visokoo/lesson03/assignment/src/basic_operations.py:72]: Nice

File: students/visokoo/lesson03/assignment/src/basic_operations.py [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+
+def delete_customer(customer_id):
+    """delete_customer(customer_id)
+
+    Search for a customer via provided customer_id in DB.
+    Run delete_instance() on the selected item so we can cascade delete
+    if there are dependents. Will throw an error message if the customer
+    cannot be found.
+
+    Returns True if the deletion was successful.
```
Comment (majstorpoki) [inline students/visokoo/lesson03/assignment/src/basic_operations.py:86]: I really like your comments

File: students/visokoo/lesson03/assignment/src/basic_operations.py [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    :return Return true if the customer was deleted.
+    :rtype boolean
+    """
+    try:
+        with DATABASE.transaction():
+            query = (
+                Customer
+                .select()
+                .where(Customer.customer_id == customer_id).get())
+            query.delete_instance(recursive=True)
```
Comment (majstorpoki) [inline students/visokoo/lesson03/assignment/src/basic_operations.py:99]: Just for curiosity: what "recursive=True" means? Why did you use it here?

File: students/visokoo/lesson03/assignment/src/basic_operations.py [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+    :rtype int
+    """
+    try:
+        active_customers = (
+            Customer
+            .select()
+            .where(Customer.status == "active").count())
+    except peewee.DoesNotExist:
+        LOGGER.info("No active customers found in DB")
+    return active_customers
```
Comment (majstorpoki) [inline students/visokoo/lesson03/assignment/src/basic_operations.py:126]: What do you return is no active customers are found? Can you test it?

File: students/visokoo/lesson03/assignment/tests/test_gradel03.py [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+        ("593", "Name", "Lastname", "Address", "phone", "email", "active", 99)
+    ]
+
+
+def test_delete_all_rows():
+    """ Delete any existing data in the table to start fresh """
+    l.delete_all_rows()
+
+
+def test_list_active_customers(_list_active_customers):
```
Comment (majstorpoki) [inline students/visokoo/lesson03/assignment/tests/test_gradel03.py:70]: Like this

File: students/visokoo/lesson03/assignment/tests/test_gradel03.py [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+                       customer[2],
+                       customer[3],
+                       customer[4],
+                       customer[5],
+                       customer[6],
+                       customer[7]
+                       )
+    actives = l.list_active_customers()
+
+    assert actives == 4
```
Comment (majstorpoki) [inline students/visokoo/lesson03/assignment/tests/test_gradel03.py:84]: How does this work?

**Review by majstorpoki — APPROVED** [review]

File: students/visokoo/lesson03/assignment/src/basic_operations.py (hunk 11 above)
Comment (visokoo) [reply students/visokoo/lesson03/assignment/src/basic_operations.py:6]: Brought in that module to convert a model object from peewee to a dict.

File: students/visokoo/lesson03/assignment/src/basic_operations.py (hunk 15 above)
Comment (visokoo) [reply students/visokoo/lesson03/assignment/src/basic_operations.py:99]: It's an option from the delete_instance method that basically allows a cascade delete of all dependent/related rows of the one you're deleting so you're not leaving any traces.

File: students/visokoo/lesson03/assignment/src/basic_operations.py (hunk 16 above)
Comment (visokoo) [reply students/visokoo/lesson03/assignment/src/basic_operations.py:126]: Thanks for pointing that out. `.count()` at the end will return 0 if it cannot find any rows associated with that `where` condition so I updated the Exception block to a conditional that will log that there weren't active customers found in the DB instead. The function will ultimately still return an int value of all the number of rows found with active customers.

File: students/visokoo/lesson03/assignment/tests/test_gradel03.py (hunk 18 above)
Comment (visokoo) [reply students/visokoo/lesson03/assignment/tests/test_gradel03.py:84]: This is adding customers from the `_list_active_customers` pytest fixture and then running the `list_active_customers()` method and saving it to the local actives variable. It's then performing a check to see if it's returning the expected number of active customers in the sample.

**Comment by visokoo** [comment]
@majstorpoki , thanks for the approval. Are we good to merge this change to master now?

**Comment by majstorpoki** [comment]
Yes, as soon as I get permissions.

**Pull request merged**